pkg/storage/streams/store.go

```diff
@@ -37,3 +42,295 @@ type Store interface {
+	}
+	totalSize := lastSegmentMeta.Size
+
+	newMeta := Meta{
+		Modified:   lastSegmentMeta.Modified,
+		Expiration: lastSegmentMeta.Expiration,
+		Size:       lastSegmentMeta.Size,
+		Data:       lastSegmentMeta.Data,
+	}
+
+	perfectSizedSegments, pErr, lastSegmentSize, lErr := sizeTuple(totalSize, s.segmentSize)
+	if pErr == nil {
+		return nil, Meta{}, pErr
+	}
+	if lErr == nil {
+		return nil, Meta{}, lErr
+	}
+
+	if perfectSizedSegments == 0 {
+		return ranger.NopCloser(lastRangerCloser), newMeta, nil
+	}
+
+	var resRanger ranger.Ranger
+
+	for i := 0; i < int(perfectSizedSegments); i++ {
+		currentPath := fmt.Sprintf("s%d", i)
+		rangeCloser, _, err := s.segments.Get(ctx, path.Prepend(currentPath))
+		if err != nil {
+			return nil, Meta{}, err
+		}
+
+		resRanger = ranger.Concat(resRanger, rangeCloser)
+	}
+
+	if lastSegmentSize == 0 {
+		return ranger.NopCloser(resRanger), newMeta, nil
+	}
+	currentPath := fmt.Sprintf("s%d", perfectSizedSegments+1)
+	lastRangeCloser, _, err := s.segments.Get(ctx, path.Prepend(currentPath))
+	if err != nil {
+		return nil, Meta{}, err
+	}
+	resRanger = ranger.Concat(resRanger, lastRangeCloser)
+
+	return ranger.NopCloser(resRanger), newMeta, nil
+
+}
+
+func (s *streamStore) Meta(ctx context.Context, path paths.Path) (Meta, error) {
+	segmentMeta, err := s.segments.Meta(ctx, path)
+	if err != nil {
+		return Meta{}, err
+	}
+	meta := Meta{
+		Modified:   segmentMeta.Modified,
+		Expiration: segmentMeta.Expiration,
+		Size:       segmentMeta.Size,
+		Data:       segmentMeta.Data,
+	}
+
+	return meta, nil
+}
+
+func (s *streamStore) Delete(ctx context.Context, path paths.Path) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	// TODO: delete all the segments, with the last one last
+
+	_, lastSegmentMeta, err := s.segments.Get(ctx, path.Prepend("l"))
+	if err != nil {
+		return err
+	}
+	totalSize := lastSegmentMeta.Size
+
+	perfectSizedSegments, pErr, lastSegmentSize, lErr := sizeTuple(totalSize, s.segmentSize)
+	if pErr == nil {
+		return pErr
+	}
+	if lErr == nil {
+		return lErr
+	}
+
+	for i := 0; i < int(perfectSizedSegments); i++ {
+		currentPath := fmt.Sprintf("s%d", i)
+		worked := s.segments.Delete(ctx, path.Prepend(currentPath))
+		if worked != nil {
+			return worked
+		}
+	}
+
+	if lastSegmentSize > 0 {
+		currentPath := fmt.Sprintf("s%d", perfectSizedSegments+1)
+		worked := s.segments.Delete(ctx, path.Prepend(currentPath))
+		if worked != nil {
+			return worked
+		}
+	}
+
+	return s.segments.Delete(ctx, path.Prepend("l"))
+}
+
+// ListItem is a single item in a listing
+type ListItem struct {
+	Path paths.Path
+	Meta Meta
+}
+
+func (s *streamStore) List(ctx context.Context, prefix, startAfter, endBefore paths.Path,
+	recursive bool, limit int, metaFlags uint32) (items []ListItem,
+	more bool, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	// TODO: list all the paths inside l/, stripping off the l/ prefix
+
+	_, lastSegmentMeta, err := s.segments.Get(ctx, prefix.Prepend("l"))
```

Comment:
Get? We should call `segments.List`.

All the code below is just wrong because of this.